src/components/swap/ConfirmSwapModal.tsx

```diff
@@ -1,39 +1,172 @@
 import { Trans } from '@lingui/macro'
-import { sendAnalyticsEvent, Trace } from '@uniswap/analytics'
-import { InterfaceModalName, SwapEventName, SwapPriceUpdateUserResponse } from '@uniswap/analytics-events'
+import { sendAnalyticsEvent, Trace, useTrace } from '@uniswap/analytics'
+import {
+  InterfaceEventName,
+  InterfaceModalName,
+  SwapEventName,
+  SwapPriceUpdateUserResponse,
+} from '@uniswap/analytics-events'
 import { Percent } from '@uniswap/sdk-core'
+import { useWeb3React } from '@web3-react/core'
+import Modal from 'components/Modal'
+import { useMaxAmountIn } from 'hooks/useMaxAmountIn'
+import { Allowance, AllowanceState } from 'hooks/usePermit2Allowance'
+import usePrevious from 'hooks/usePrevious'
 import { getPriceUpdateBasisPoints } from 'lib/utils/analytics'
-import { ReactNode, useCallback, useEffect, useMemo, useState } from 'react'
+import { ReactNode, useCallback, useEffect, useState } from 'react'
 import { InterfaceTrade } from 'state/routing/types'
+import { useIsTransactionConfirmed } from 'state/transactions/hooks'
+import invariant from 'tiny-invariant'
 import { formatSwapPriceUpdatedEventProperties } from 'utils/loggingFormatters'
+import { didUserReject } from 'utils/swapErrorToUserReadableMessage'
 import { tradeMeaningfullyDiffers } from 'utils/tradeMeaningFullyDiffer'
 
-import TransactionConfirmationModal, {
-  ConfirmationModalContent,
-  TransactionErrorContent,
-} from '../TransactionConfirmationModal'
+import { ConfirmationModalContent } from '../TransactionConfirmationModal'
+import {
+  ErrorModalContent,
+  PendingConfirmModalState,
+  PendingModalContent,
+  PendingModalError,
+} from './PendingModalContent'
 import SwapModalFooter from './SwapModalFooter'
 import SwapModalHeader from './SwapModalHeader'
 
+export enum ConfirmModalState {
+  REVIEWING,
+  APPROVING_TOKEN,
+  PERMITTING,
+  PENDING_CONFIRMATION,
+}
+
+function isAllowing(confirmModalState: ConfirmModalState) {
+  return confirmModalState === ConfirmModalState.APPROVING_TOKEN || confirmModalState === ConfirmModalState.PERMITTING
+}
+
+function useConfirmModalState({
+  trade,
+  allowedSlippage,
+  onSwap,
+  allowance,
+  doesTradeDiffer,
+}: {
+  trade: InterfaceTrade
+  allowedSlippage: Percent
+  onSwap: () => void
+  allowance: Allowance
+  doesTradeDiffer: boolean
+}) {
+  const [confirmModalState, setConfirmModalState] = useState<ConfirmModalState>(ConfirmModalState.REVIEWING)
+  const [approvalError, setApprovalError] = useState<PendingModalError>()
+  const [pendingModalSteps, setPendingModalSteps] = useState<PendingConfirmModalState[]>([])
+
+  const prepareSwapFlow = useCallback(() => {
```

Comment:
I don't think this needs to be a function. You could make it an array with `useMemo` and it would declaratively update itself when `allowance` is set. That way, it is not required to set it imperatively.

Reply:
this actually needs to be a function, because we don't want the steps to update when you're in the middle of the flow. if the flow starts with 3 steps, it shouldn't lose the first 2 step indicators once the `allowance` is set

Reply:
I mean you could always make it `useMemo` instead of a function - right? That way it would only run every time allowance changes, no need to call it.

Reply:
In other words, why not just make it:
```ts
const steps = useMemo(() => {
 // do the thing
}, [allowance])
```

Reply:
@grabbou we don't want it to run when the allowance changes though, only when the user starts the flow. 

when the allowance changes, we move from the approval phase to the final step, but we don't want the number of steps indicated at the bottom of the modal to change (it should still be two or three depending on what it was at the start)

Reply:
discussed offline

Reply:
@just-toby you can accomplish that by setting a variable that only reads from the initial value of `allowance`

```
const initialAllowance = useRef()

useEffect(() => {
   if (initialAllowance) return 

  initialAllowance.current = allowance
}, [allowance])

// and then use initialAllowance only
```

Reply:
bump on this comment ^